packages/trace-viewer/src/ui/uiModeView.tsx

```diff
@@ -83,62 +79,129 @@ export const UIModeView: React.FC<{}> = ({
-    (async () => {
-      const status = await connection.runGlobalSetup();
-      if (status === 'passed')
-        await refreshRootSuite(connection);
-      setIsLoading(false);
-      const { hasBrowsers } = await connection.checkBrowsers();
-      setHasBrowsers(hasBrowsers);
-    })();
+    setTestServerConnection(new TestServerConnection(wsURL.toString()));
   }, []);
 
+  // Load tests on startup.
+  React.useEffect(() => {
+    inputRef.current?.focus();
+    setIsLoading(true);
+    reloadTests();
+  }, [reloadTests]);
+
+  // Wire server connection to the auxiliary UI features.
   React.useEffect(() => {
     if (!testServerConnection)
       return;
     const disposables = [
-      ...wireConnectionListeners(testServerConnection),
+      testServerConnection.onStdio(params => {
+        if (params.buffer) {
+          const data = atob(params.buffer);
+          xtermDataSource.write(data);
+        } else {
+          xtermDataSource.write(params.text!);
+        }
+      }),
       testServerConnection.onClose(() => setIsDisconnected(true))
     ];
+    xtermDataSource.resize = (cols, rows) => {
+      xtermSize = { cols, rows };
+      testServerConnection.resizeTerminalNoReply({ cols, rows });
+    };
     return () => {
       for (const disposable of disposables)
         disposable.dispose();
     };
   }, [testServerConnection]);
 
+
+  // This is the main routine, every time connection updates it starts the
+  // whole workflow.
   React.useEffect(() => {
-    inputRef.current?.focus();
+    if (!testServerConnection)
+      return;
+
+    let throttleTimer: NodeJS.Timeout | undefined;
+    const teleSuiteUpdater = new TeleSuiteUpdater({
+      onUpdate: immediate => {
+        clearTimeout(throttleTimer);
+        throttleTimer = undefined;
+        if (immediate) {
+          setTestModel(teleSuiteUpdater.asModel());
+        } else if (!throttleTimer) {
+          throttleTimer = setTimeout(() => {
+            setTestModel(teleSuiteUpdater.asModel());
+          }, 250);
+        }
+      },
+      onError: error => {
+        xtermDataSource.write((error.stack || error.value || '') + '\n');
+      },
+      pathSeparator,
+    });
+
+    const updateList = async () => {
+      setIsLoading(true);
+      const result = await testServerConnection.listTests({});
+      teleSuiteUpdater.processListReport(result.report);
+      setIsLoading(false);
+    };
+
+    setTestModel(undefined);
     setIsLoading(true);
-    reloadTests();
-  }, [reloadTests]);
+    setWatchedTreeIds({ value: new Set() });
+    (async () => {
+      const status = await testServerConnection.runGlobalSetup();
+      if (status !== 'passed')
+        return;
+      const result = await testServerConnection.listTests({});
+      teleSuiteUpdater.processListReport(result.report);
+
+      testServerConnection.onListChanged(updateList);
+      testServerConnection.onReport(params => {
+        teleSuiteUpdater.processTestReport(params);
```

Comment:
nit: perhaps `processTestReportEvent` would be better, to differentiate with `processListReport`?